runtime/extern.c

```diff
@@ -132,68 +151,145 @@ static struct extern_item * extern_resize_stack(struct extern_item * sp)
   return newstack + sp_offset;
 }
 
-/* Initialize the trail */
+/* Multiplicative Fibonacci hashing
+   (Knuth, TAOCP vol 3, section 6.4, page 518).
+   HASH_FACTOR is (sqrt(5) - 1) / 2 * 2^wordsize. */
+#ifdef ARCH_SIXTYFOUR
+#define HASH_FACTOR 11400714819323198486UL
+#else
+#define HASH_FACTOR 2654435769UL
+#endif
+#define Hash(v) (((uintnat)(v) * HASH_FACTOR) >> pos_table.shift)
+
+/* When the table becomes 2/3 full, its size is increased. */
+#define Threshold(sz) (((sz) * 2) / 3)
+
+/* Initialize the position table */
 
-static void init_extern_trail(void)
+static void extern_init_position_table(void)
 {
-  extern_trail_block = &extern_trail_first;
-  extern_trail_cur = extern_trail_block->entries;
-  extern_trail_limit = extern_trail_block->entries + ENTRIES_PER_TRAIL_BLOCK;
+  if (extern_flags & NO_SHARING) return;
+  pos_table.size = POS_TABLE_INIT_SIZE;
+  pos_table.shift = 8 * sizeof(value) - POS_TABLE_INIT_SIZE_LOG2;
+  pos_table.mask = POS_TABLE_INIT_SIZE - 1;
+  pos_table.threshold = Threshold(POS_TABLE_INIT_SIZE);
+  pos_table.present = pos_table_present_init;
+  pos_table.entries = pos_table_entries_init;
+  memset(pos_table_present_init, 0, sizeof(pos_table_present_init));
 }
 
-/* Replay the trail, undoing the in-place modifications
-   performed on objects */
+/* Free the position table */
 
-static void extern_replay_trail(void)
+static void extern_free_position_table(void)
 {
-  struct trail_block * blk, * prevblk;
-  struct trail_entry * ent, * lim;
+  if (pos_table.present != pos_table_present_init) {
+    caml_stat_free(pos_table.present);
+    caml_stat_free(pos_table.entries);
+    pos_table.present = pos_table_present_init;
```

Comment:
The purpose of this reinitialization of `pos_table.present` is to make `extern_free_position_table` "idempotent": if it is called  a second time, nothing will happen, especially not a double-free error.  (I'm pretty sure that the current implementation calls `extern_free_stack` and `extern_free_position_table` twice in most cases... something to address later!)

I'll add a comment to clarify the intent.

We could reset `pos_table.entries` but it's not necessary for this purpose, and it will be redundant with the initialization done in `extern_init_position_table`.

We could move some of the initialization done in  `extern_init_position_table` to `extern_free_position_table`, with the same "reinitialize the globals for next time around" logic that we use in `extern_free_stack`, but the initialization of the position table is more complex than that of the stack, so I'd rather keep it all in one place.

Reply:
Thanks for the explanation.